### PR TITLE
[WIP] Modify vim_strbyte() to take advantage of strchr() (again)

### PR DESCRIPTION
**Pull request by amaro**
Apologies, I closed the previous PR #2500 unintentionally. I modified the small commit as requested.

@justinmk "Besides unit tests, some reproducible performance results with actual vimscript (or searching via / or search()) is a large part of the work needed for #1476."

Any pointers on how to get the performance results?

ref #1476

### CODE REVIEW

**Comment by Pyrohh** [comment]
> Any pointers on how to get the performance results?

I recommend looking into `gprof`. You might already have it installed if on Linux, as it's part of [Gnu Binutils](https://www.gnu.org/software/binutils/).

**Comment by amaro** [comment]
Right. I understand the basics of using a profiler. My question is more geared towards how to run a particular nvim execution that, say, searches for N number of strings in a given file?

**Comment by Pyrohh** [comment]
It looks like benchmarking support was added quite recently in #2295 (just run `make benchmark`). 

[/test/benchmark/bench_re_freeze_spec.lua](https://github.com/neovim/neovim/blob/master/test/benchmark/bench_re_freeze_spec.lua) might be of use.

**Comment by amaro** [comment]
I ran the benchmarks 10 times for master and this branch, with -O2 builds. These are the results:

Master:
bench 1 avg: 0.0206145 secs
bench 2 avg: 0.0170913 secs
bench 3 avg: 12.3356079 secs

Modified branch:
bench 1 avg: 0.020888 secs
bench 2 avg: 0.0174081 secs
bench 3 avg: 12.347608 secs

The modified branch is:
0.09% slower in bench1
1.85% slower in bench2
1.32% slower in bench3

Thus, I don't know if it makes sense anymore to make this change.

**Comment by Pyrohh** [comment]
How are you benchmarking it? Could you push your test? 

Also what libc are you using? And are you compiling with `-flto` (link time optimization, see [this patch](https://gist.githubusercontent.com/Pyrohh/7b2e1bf44f6336cdcd4d/raw/2f27c61b0b586ac6db1654885cca478500a818cd/-))?

**Comment by amaro** [comment]
Maybe I misunderstood. I glanced at the available benchmarks and it seems like its doing some regex stuff. Thus, I thought that those benchmarks would hit the function I am modifying. So I didn't write any new tests.

I'll try the -flto thing.

**Comment by Pyrohh** [comment]
Set back to WIP for now, feel free to change it to RFC once you've updated it.

File: src/nvim/strings.c [hunk 1]
```diff
@@ -412,22 +412,18 @@ char_u *vim_strchr(const char_u *string, int c)
...
+char_u *vim_strbyte(const char_u *buf, int c)
   FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
 {
-  const char_u *p = string;
-
-  while (*p != NUL) {
-    if (*p == c)
-      return (char_u *) p;
-    ++p;
+  if (c > (unsigned char) -1 || c <= 0) {
```
Comment (splinterofchaos) [inline src/nvim/strings.c]: Doesn't this conditional return true for any value?

Comment (oni-link) [inline src/nvim/strings.c]: `(unsigned char) -1 == 255`

Comment (splinterofchaos) [inline src/nvim/strings.c]: Right, I was confused by the cast. What about `UCHAR_MAX`?

**Comment by splinterofchaos** [comment]
I think to see any performance improvements, we'd have to more this function to the header and label as `inline`. That could potentially enable minor compiler optimizations, but still not likely noticeable one because vim does so much other work that searching a string can't be a bottleneck. 

I think we should use `strchr` over `vim_strbyte`, but not for performance reasons.

**Comment by amaro** [comment]
I took 4 measurements using the "make benchmark" tests available. The following summary only contains the results for the benchmark 3, although I have the results for the other 2 if needed.

Measurement                   Seconds             Speedup
1. -O2 master	                 11.78447927	    1
2. -O2 newbranch	         11.78921025	    0.9995
3. -flto -O2 master	         17.90785878	    0.6580
4. -flto -O2 newbranch	 17.95858	    0.6562

The baseline for the speedup is result 1, -O2 master branch.

As you can see, -flto actually hurts. And I believe the performance loss for the change (result 2) is negligible.

The complete measurements are available here: https://docs.google.com/spreadsheets/d/1rIW3PlcZ351DgNnqxYnsdx0aG8poc0OSG8O49i2Qz2Q

**Comment by justinmk** [comment]
Hmm. As I tried to imply, a necessary task for #1476 is to *find a code path for which `vim_strbyte` is a hotspot*. Doesn't seem like `make benchmark` is exercising that.

**Comment by amaro** [comment]
Cool, I'll work on that then. But this also proves that -flto is not helping here, as it was implied before.

**Comment by Pyrohh** [comment]
> Hmm. As I tried to imply, a necessary task for #1476 is to find a code path for which vim_strbyte is a hotspot. Doesn't seem like make benchmark is exercising that.

This seems like the issue. Anyways, `-flto` should not have such a significant negative impact on performance; perhaps you're running an older version of GCC? The 4.8 and 4.9 series saw significant improvements in compilation time (but also in runtime performance).

**Comment by amaro** [comment]
Yes. Initially I tried with gcc 4.4.7, but I just tried with gcc 4.8.2 and saw no significant difference. Can you reference the gcc's improvements you are referring to?

**Comment by Pyrohh** [comment]
> Can you reference the gcc's improvements you are referring to?

My suggestion came about because of these two articles:


- http://hubicka.blogspot.com/2014/04/linktime-optimization-in-gcc-1-brief.html -- This one is a fairly high level overview of recent improvements and various history:
- http://hubicka.blogspot.com/2014/04/linktime-optimization-in-gcc-2-firefox.html -- This one has a lot of benchmarking, and various comparisons between recent major GCC versions (excluding 5.0):

Anyways, my suggestion to use LTO was a complete guess, so... eh.

**Comment by Pyrohh** [comment]
Since there hasn't been much progress I'm going to close this for now, but please feel free to reopen if you want to work on this more @amaro.

**Pull request closed without merging**